execinfra,colflow: improve and harden MetadataSource interface

yuzefovich opened the pull request:
**execinfra: remove ctx argument from DrainMeta**

This commit updates the contract of `DrainMeta` method to use the
context owned by the MetadataSource (the one it received in `Start` or
`Init`) because it is a lot less confusing this way.

Release note: None

**colflow: only drain sources if they were properly initialized**

Previously, in several cases we could have called `DrainMeta` on some
operators even if those weren't properly initialized. Such behavior
should not be allowed, and this commit puts the protection in place in
several "root" components (outboxes, hash routers, parallel unordered
synchronizers; the only other "root" component materializer already had
similar protection in place). The follow up commit will clean up the
MetadataSource interface.

Such misbehavior was never intended to be allowed but was safely ignored
because up until recently the uninitialized operators (at least in most
cases) were able to handle DrainMeta call gracefully. This changed when
we began the effort to centralize the context management.

Fixes: #64288.
Fixes: #64362.
Fixes: #64363.
Fixes: #64364.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61323)
<!-- Reviewable:end -->

yuzefovich committed:
execinfra: remove ctx argument from DrainMeta

This commit updates the contract of `DrainMeta` method to use the
context owned by the MetadataSource (the one it received in `Start` or
`Init`) because it is a lot less confusing this way.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61323)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/colfetcher/colbatch_scan.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-MZ1Cvg0-9lrkmMkv73t:-MZ1Cvg0-9lrkmMkv73u:b-7f0oz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d830f919471ad7d12924c8366027917491ea549d/pkg/sql/colfetcher/colbatch_scan.go#L130)):*
> ```Go
> 	meta.Metrics.RowsRead = s.GetRowsRead()
> 	trailingMeta = append(trailingMeta, *meta)
> 	if s.tracingSpan != nil {
> ```

I was wrong - we can, in fact, remove this check. The reasoning is that if tracing is not enabled, then `tracingSpan` is nil but also `GetTraceData` will return `nil`; if the tracing is enabled, then we must have instantiated a tracing span in `Init`, and we must collect the trace.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colflow: only drain sources if they were properly initialized

Previously, in several cases we could have called `DrainMeta` on some
operators even if those weren't properly initialized. Such behavior
should not be allowed, and this commit puts the protection in place in
several "root" components (outboxes, hash routers, parallel unordered
synchronizers; the only other "root" component materializer already had
similar protection in place). The follow up commit will clean up the
MetadataSource interface.

Such misbehavior was never intended to be allowed but was safely ignored
because up until recently the uninitialized operators (at least in most
cases) were able to handle DrainMeta call gracefully. This changed when
we began the effort to centralize the context management.

Release note: None

yuzefovich commented:
RFAL.

yuzefovich commented:
@rytaft friendly ping on this - recent sqlsmith issues are fixed by the second commit here.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for the ping!

> Reviewed 15 of 25 files at r1, 10 of 10 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-:-MZSWjzT7NR5JQZJPBKA:b-8oyr3x)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexecop/operator.go, line 365 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-MZSWPEu1yzyHJca80ll:-MZSWPEu1yzyHJca80lm:bgt7nbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb4c89f5de80492abc6fa6151dd310c4689cfd8/pkg/sql/colexecop/operator.go#L365)):*
> ```Go
> // operators that can produce metadata.
> // TODO(yuzefovich): remove this interface in favor of DrainableOperator and
> // clarify that calling DrainMeta on uninitialized operator is illegal.
> ```

nit: on uninitialized operator -> on an uninitialized operator
___
*[pkg/sql/colfetcher/colbatch_scan.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-MZ1Cvg0-9lrkmMkv73t:-MZSV6ST0iolBDJkea5V:bwuhsja) ([raw file](https://github.com/cockroachdb/cockroach/blob/d830f919471ad7d12924c8366027917491ea549d/pkg/sql/colfetcher/colbatch_scan.go#L130)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I was wrong - we can, in fact, remove this check. The reasoning is that if tracing is not enabled, then `tracingSpan` is nil but also `GetTraceData` will return `nil`; if the tracing is enabled, then we must have instantiated a tracing span in `Init`, and we must collect the trace.
</blockquote></details>

Is this the same reasoning for the check you deleted above for `ShouldCollectStats` (I think I missed some context here)?
___
*[pkg/sql/rowexec/tablereader.go, line 304 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-MZSVko87cfV9XzofD2z:-MZSVko90LdgP7mq1dEg:b-8uxdhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/41b577b05ca23ffc668e161ec4336d57dc834f67/pkg/sql/rowexec/tablereader.go#L304)):*
> ```Go
> }
> 
> func (tr *tableReader) generateMeta(ctx context.Context) []execinfrapb.ProducerMetadata {
> ```

should you remove ctx here since it's no longer used?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-:-MZT3a3V0Xb9HgLM3uhO:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/colfetcher/colbatch_scan.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-MZ1Cvg0-9lrkmMkv73t:-MZT2Evp8_F6oXfJQWYv:bkrjb3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d830f919471ad7d12924c8366027917491ea549d/pkg/sql/colfetcher/colbatch_scan.go#L130)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is this the same reasoning for the check you deleted above for `ShouldCollectStats` (I think I missed some context here)?
</blockquote></details>

Yeah, pretty much. I updated the comment there. The deletion is not exactly equivalent, but I think it makes sense to start a new tracing span if the tracing is enabled (the difference with that change is we will now start a span if the tracing is enabled but the stats weren't explicitly requested to be collected).

The change here (about `GetTraceData`) and there (about `ProcessorSpan`) kinda rely on each other.
___
*[pkg/sql/rowexec/tablereader.go, line 304 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61323#-MZSVko87cfV9XzofD2z:-MZT314h5Sitx8rYMoMp:befvztr) ([raw file](https://github.com/cockroachdb/cockroach/blob/41b577b05ca23ffc668e161ec4336d57dc834f67/pkg/sql/rowexec/tablereader.go#L304)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should you remove ctx here since it's no longer used?
</blockquote></details>

Yeah, good point, done.


<!-- Sent from Reviewable.io -->
